Add basic HTTP health check

richlander opened the pull request:
I followed these instructions: https://learn.microsoft.com/aspnet/core/host-and-deploy/health-checks.

I tested with `curl` and Kubernetes. Both worked.

```bash
% curl http://localhost:8000/healthz
Healthy
```

mthalman reviewed (commented):
In my mind, having the app capable of doing a health check and leaving it at that, it doesn't go far enough. We should also demonstrate how it's used in the Dockerfile. I think we should update the Dockerfiles to include a `HEALTHCHECK` instruction. This is still useful in plain old Docker, outside the context of Kubernetes. The health check status will show up in a `docker ps` output of the running sample container.

richlander commented:
I will do that Matt. That's a good idea.

richlander commented:
Before I go any further, I want to validate the change for just `Dockerfile`. I'm happy to make the same change to the others.

Before:

```bash
$ docker run --rm -d -p 8000:80 mcr.microsoft.com/dotnet/samples:aspnetapp
42f412fdb600bbdc515ec7839f2632bdb3658400f76e021b1da83b61544f8033
$ docker ps -l
CONTAINER ID   IMAGE                                        COMMAND         CREATED         STATUS         PORTS                  NAMES
42f412fdb600   mcr.microsoft.com/dotnet/samples:aspnetapp   "./aspnetapp"   4 seconds ago   Up 3 seconds   0.0.0.0:8000->80/tcp   elegant_buck
```

After:

```bash
$ docker run --rm -d -p 8000:80 aspnetapp
5a992022ff999d10e82992b7f729105ba12ef630655a3a5ccf0f76424d61f6bc
$ docker ps -l
CONTAINER ID   IMAGE       COMMAND                  CREATED          STATUS                             PORTS
       NAMES
5a992022ff99   aspnetapp   "dotnet aspnetapp.dll"   10 seconds ago   Up 10 seconds (health: starting)   0.0.0.0:8000->80/tcp   beautiful_kilby
docker ps -l
CONTAINER ID   IMAGE       COMMAND                  CREATED          STATUS                    PORTS                  NAMES
5a992022ff99   aspnetapp   "dotnet aspnetapp.dll"   34 seconds ago   Up 33 seconds (healthy)   0.0.0.0:8000->80/tcp   beautiful_kilby
```

Looks good?

jander-msft commented on samples/aspnetapp/Dockerfile:
```diff
@@ -12,6 +12,10 @@ RUN dotnet publish -c Release -o /app --use-current-runtime --self-contained fal
+RUN apt-get update \ 
+    && apt-get install -y --no-install-recommends curl \
+    && rm -rf /var/lib/apt/lists/*
+HEALTHCHECK CMD curl http://localhost/healthz
```
Does this need to somehow account for if the user specifies HTTPS_PORTS OR HTTPS_PORTS environment variables on the container?

richlander replied:
That's a great point. Ports suck.

With .NET 8, that will be easier. This is still .NET 7. I don't what to do. I added a comment.

richlander commented on samples/aspnetapp/Dockerfile:
```diff
@@ -12,6 +12,11 @@ RUN dotnet publish -c Release -o /app --use-current-runtime --self-contained fal
+RUN apt-get update \ 
+    && apt-get install -y --no-install-recommends curl \
+    && rm -rf /var/lib/apt/lists/*
+# This port needs to map the port being used
```
```suggestion
# This port needs to match the port being used
```

MichaelSimons commented on samples/aspnetapp/Dockerfile:
```diff
@@ -12,6 +12,11 @@ RUN dotnet publish -c Release -o /app --use-current-runtime --self-contained fal
 
 # final stage/image
 FROM mcr.microsoft.com/dotnet/aspnet:7.0
+RUN apt-get update \ 
```
This is intended to be a multi-platform Dockerfile.  This breaks that contract.  This is another case for needing https://github.com/dotnet/dotnet-docker/issues/4300.

richlander replied:
I also think that this approach is ugly and we should provide a helper app. My intent was to demonstrate the basic case and see where the conversation led.

mthalman replied:
> I also think that this approach is ugly and we should provide a helper app.

Even with the helper app, I'm wondering whether this is going to overcomplicate the sample and it should be factored out into a separate Dockerfile that demonstrates the health check.

MichaelSimons commented on samples/aspnetapp/Dockerfile:
```diff
@@ -12,6 +12,11 @@ RUN dotnet publish -c Release -o /app --use-current-runtime --self-contained fal
+    && apt-get install -y --no-install-recommends curl \
+    && rm -rf /var/lib/apt/lists/*
+# This port needs to match the port being used
+HEALTHCHECK CMD curl http://localhost:80/healthz
```
When I first saw this, I thought it was a bit weird to see the HEALTHCHECK instruction prior to the app being copied into the image.  After thinking about it, it makes a bit of sense in that it never changes and therefore can better utilize the docker layer cache in a build environment.  That aside I wonder if it will be confusing to the reader.

richlander replied:
I think this approach is the correct one. The caching doesn't add much but it is caching all the same. The `ASPNET_CORE_URLS` ENV is added before all of this. It's just layering of one form or another.

richlander commented:
We could do the healthcheck in a separate Dockerfile. However, we'd want to push that version to MCR.

richlander commented:
I switched to Alpine (because `wget` is there) and added a new pattern that could be used to collapse other Alpine Dockerfiles. I'm happy to do that in another PR.

In our initial conversations, we were going to call this one `Dockerfile.health`. It would be best to avoid that if we have the goal of enable health checks throughout our fleet of samples.

I will add this to the manifest.json file once everyone is happy with the approach. I tested it and it works as expected.

mthalman reviewed (commented):
I think these file names should indicate their difference from the result of the samples. How about `Dockerfile.healthcheck-alpine-slim` and `Dockerfile.healthcheck-nanoserver-slim`?

The https://github.com/dotnet/dotnet-docker/blob/main/manifest.samples.json file should be updated to reference these Dockerfiles so that they will be published.

Did we also want to update https://github.com/dotnet/dotnet-docker/blob/main/samples/aspnetapp/README.md to have a blurb that pointed people to these Dockerfiles and explained the health check?

MichaelSimons commented on samples/aspnetapp/Dockerfile.alpine-slim:
```diff
@@ -0,0 +1,29 @@
+FROM mcr.microsoft.com/dotnet/runtime-deps:7.0-alpine
+WORKDIR /app
+COPY --from=build /app .
+# This port needs to match the port being used
```
IMO this would benefit by having more whitespace.  Specifically an empty line before this comment would help make it standout more.

richlander replied:
```suggestion

# This port needs to match the port being used
```

richlander commented:
On the naming, it's a question of where we are headed with this. I chose the naming w/o healthcheck with the idea that:

- These new Dockerfiles would replace the existing slim ones (for Alpine and Nanoserver to start).
- It would be odd to not have the health check in the Alpine and Nanoserver ones given that it is just one more (useful) line.
- We'll design an approach for the rest of the dockerfiles in a bit.
- We can point out that we have healthcheck examples from other docs/READMEs.

richlander commented:
You can check the output of health checks via `docker inspect`. How you use `curl` and `wget` matters for that. I also discovered that the details of using these tools differs a fair bit.

On Alpine, the following appears to be the best approach: `HEALTHCHECK CMD wget -qO- -t1 http://localhost:80/healthz`.

It produces the following output for the success case:

$ docker inspect 21ec64322542 | jq .[-1].State.Health
{
  "Status": "healthy",
  "FailingStreak": 0,
  "Log": [
    {
      "Start": "2023-01-26T23:18:42.790331305Z",
      "End": "2023-01-26T23:18:42.953258349Z",
      "ExitCode": 0,
      "Output": "Healthy"
    },
    {
      "Start": "2023-01-26T23:19:12.962127063Z",
      "End": "2023-01-26T23:19:13.095388607Z",
      "ExitCode": 0,
      "Output": "Healthy"
    }
  ]
}

And for the failure case:

```bash
$ docker inspect 0fd5206a0f1b | jq .[-1].State.Health
{
  "Status": "starting",
  "FailingStreak": 2,
  "Log": [
    {
      "Start": "2023-01-26T23:25:14.920091363Z",
      "End": "2023-01-26T23:25:15.090543745Z",
      "ExitCode": 1,
      "Output": "wget: server returned error: HTTP/1.1 404 Not Found\n"
    },
    {
      "Start": "2023-01-26T23:25:45.098815567Z",
      "End": "2023-01-26T23:25:45.168346131Z",
      "ExitCode": 1,
      "Output": "wget: server returned error: HTTP/1.1 404 Not Found\n"
    }
  ]
}
```

I found that Alpine and Ubuntu had different behavior, in terms of the silent behavior while still printing error information. I ended up with the following.

- On Alpine, use `-q`.
-  On Ubuntu, use `-nv`.

Works?

mthalman commented on samples/aspnetapp/README.md:
````diff
@@ -45,35 +62,49 @@ We recommend that you do not use `--rm` in production. It cleans up container re
+You can also look at health status with `docker inspect`. The following pattern uses `jq`, which makes it much easier to drill in on the interesting data.
+
+```bash
+$ docker inspect b143cf4ac0d1 | jq .[-1].State.Health
````
Consider providing a PowerShell version of this as well.

richlander replied:
Ah, you caught me being lazy. Gah!

mthalman reviewed (commented):
You'll need to also run `eng/readme-templates/Get-GeneratedReadmes.ps1` to update the readmes due to the Dockerfile path change in the manifest.

Saibamen commented:
This feature will be enabled in base Docker images?
It could be enabled by setting ENV variable (to make sure that Developer enables Healthz on code-level).

richlander commented:
That occurred to me, but it doesn't sound like the best model. Setting an ENV and opting into `HEALTHCHECK` seems pretty similar. We want to ensure that all the capabilities are in place to use `HEALTHCHECK`. It's 100% possible today, just less pretty and convenient than we realized.

richlander commented on samples/aspnetapp/README.md:
````diff
@@ -45,35 +62,64 @@ We recommend that you do not use `--rm` in production. It cleans up container re
 ```
 
-## Build an image for Windows Server Core
+After a while, you will see that the app is no longer "starting".
````
```suggestion
After 30s, the status should transition to "healthy" from "health: starting".
```

richlander commented:
Can I do args for the nanoserver Dockerfile? Ideally, we could avoid duplicating the Dockerfile just to accommodate the difference between the two FROMs.

mthalman commented:
> Can I do args for the nanoserver Dockerfile? Ideally, we could avoid duplicating the Dockerfile just to accommodate the difference between the two FROMs.

Can you explain with an example? What are you looking to reuse that you'd use ARGs for?

richlander commented:
`Docker.nanoserver-slim`, for multiple versions of nanoserver. Right now, it is targeting just one and I'm not sure how to resolve that.

Saibamen commented:
> Is this a capability that you will use? Have you been using healthcheck to date? If we only implement it in these samples, will you apply that pattern to your Dockerfiles?

No, I was not using healthchecks before, but I think I will be using it for Docker images where we don't have implemented `"api/version" controller's action

mthalman commented:
> `Docker.nanoserver-slim`, for multiple versions of nanoserver. Right now, it is targeting just one and I'm not sure how to resolve that.

Ah, got it. I don't have a problem using an ARG for that, as long as it gets defaulted.

richlander commented:
I was super unclear as to what I was asking. How do I specify an arg in the manifests.json file? I didn't see that used?

mthalman commented:
> I was super unclear as to what I was asking. How do I specify an arg in the manifests.json file? I didn't see that used?

Here's an example: https://github.com/dotnet/docker-tools/blob/bf5e6ad6053b16d023b4021bef892403783709d2/src/Microsoft.DotNet.ImageBuilder/manifest.json#L28
